File: app/rl/strategies/token_bucket.py
import time
import math
from pathlib import Path
from typing import Dict, Any, Tuple
from redis.asyncio import Redis
from app.rl.keys import bucket_key
from app.rl.schemas import CheckDecision

_SCRIPT = None
_SCRIPT_TEXT = None


def _get_script_text() -> str:
    global _SCRIPT_TEXT
    if _SCRIPT_TEXT is None:
        path = Path(__file__).resolve().parent.parent / "scripts" / "token_bucket.lua"
        _SCRIPT_TEXT = path.read_text(encoding="utf-8")
    return _SCRIPT_TEXT


async def _eval_script(redis: Redis, keys: list[str], args: list):
    script_text = _get_script_text()
    # Special-case fakeredis which often lacks evalsha/register_script
    if "fakeredis" in type(redis).__module__:
        return await redis.eval(script_text, len(keys), *keys, *args)
    # Prefer register_script if available
    reg = getattr(redis, "register_script", None)
    if callable(reg):
        script = reg(script_text)
        return await script(keys=keys, args=args)
    # Fallback to evalsha if supported
    load = getattr(redis, "script_load", None)
    evalsha = getattr(redis, "evalsha", None)
    if callable(load) and callable(evalsha):
        sha = await load(script_text)
        return await evalsha(sha, len(keys), *keys, *args)
    # Last resort: EVAL every time
    return await redis.eval(script_text, len(keys), *keys, *args)


async def check(
    redis: Redis,
    key: str,
    *,
    capacity: int,
    refill_rate_per_sec: float,
    cost: int = 1,
    now_ms: int | None = None,
) -> CheckDecision:
    if now_ms is None:
        now_ms = int(time.time() * 1000)
    # Fallback pure-Python path for fakeredis (no Lua support)
    if "fakeredis" in type(redis).__module__:
        data = await redis.hgetall(key)
        tokens = float(data.get("tokens", capacity) if data else capacity)
        ts = int(data.get("ts", now_ms) if data else now_ms)
        elapsed_ms = max(0, now_ms - ts)
        tokens = min(capacity, tokens + (elapsed_ms / 1000.0) * refill_rate_per_sec)
        allowed = tokens >= cost
        if allowed:
            tokens -= cost
        retry_after_ms = 0
        if not allowed and refill_rate_per_sec > 0:
            missing = max(0.0, cost - tokens)
            retry_after_ms = int((missing / refill_rate_per_sec) * 1000.0 + 0.5)
        await redis.hset(key, mapping={"tokens": tokens, "ts": now_ms})
        ttl_sec = (
            int((capacity / refill_rate_per_sec) + 5)
            if refill_rate_per_sec > 0
            else 3600
        )
        await redis.expire(key, ttl_sec)
        limit = int(capacity)
        remaining = int(tokens)
        reset_at_s = math.ceil((now_ms + retry_after_ms) / 1000)
        headers = {
            "X-RateLimit-Limit": str(limit),
            "X-RateLimit-Remaining": str(remaining),
            "X-RateLimit-Reset": str(reset_at_s),
            "Retry-After": str(math.ceil(retry_after_ms / 1000)),
        }
        return CheckDecision(
            allowed=allowed,
            remaining=remaining,
            limit=limit,
            reset_at=reset_at_s,
            retry_after_ms=retry_after_ms,
            algorithm="token_bucket",
            headers=headers,
        )

    res = await _eval_script(
        redis, keys=[key], args=[capacity, refill_rate_per_sec, now_ms, cost]
    )
    # res: [allowed, remaining, capacity, retry_after_ms]
    allowed = int(res[0]) == 1
    remaining = int(res[1])
    limit = int(res[2])
    retry_after_ms = int(res[3])
    reset_at_s = math.ceil((now_ms + retry_after_ms) / 1000)
    headers = {
        "X-RateLimit-Limit": str(limit),
        "X-RateLimit-Remaining": str(remaining),
        "X-RateLimit-Reset": str(reset_at_s),
        "Retry-After": str(math.ceil(retry_after_ms / 1000)),
    }
    return CheckDecision(
        allowed=allowed,
        remaining=remaining,
        limit=limit,
        reset_at=reset_at_s,
        retry_after_ms=retry_after_ms,
        algorithm="token_bucket",
        headers=headers,
    )


# Backward compatible wrapper used by current engine
async def token_bucket_check(
    redis: Redis,
    api_key: str,
    namespace: str,
    subject: str,
    params: Dict[str, Any],
    cost: int,
) -> Tuple[bool, int, int]:
    capacity = int(params.get("capacity", 10))
    refill_rate = float(params.get("refill_rate", 5))
    name = params.get("name", "default")
    key = bucket_key(namespace, subject, f"tb:{name}")
    decision = await check(
        redis,
        key,
        capacity=capacity,
        refill_rate_per_sec=refill_rate,
        cost=cost,
    )
    return decision.allowed, decision.remaining, decision.retry_after_ms

File: app/rl/scripts/token_bucket.lua
-- Atomic Token Bucket using Redis hash
-- KEYS[1] = bucket key
-- ARGV = [capacity, refill_rate_per_sec, now_ms, cost]

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill = tonumber(ARGV[2])
local now_ms = tonumber(ARGV[3])
local cost = tonumber(ARGV[4])

-- Read current state
local data = redis.call('HMGET', key, 'tokens', 'ts')
local tokens = tonumber(data[1])
local ts = tonumber(data[2])

if tokens == nil then tokens = capacity end
if ts == nil then ts = now_ms end

-- Refill tokens
local elapsed_ms = now_ms - ts
if elapsed_ms < 0 then elapsed_ms = 0 end
local elapsed_sec = elapsed_ms / 1000.0
local new_tokens = tokens
if refill and refill > 0 then
  new_tokens = math.min(capacity, tokens + elapsed_sec * refill)
else
  new_tokens = math.min(capacity, tokens)
end

-- Decision
local allowed = 0
if new_tokens >= cost then
  allowed = 1
  new_tokens = new_tokens - cost
end

-- Compute retry_after_ms
local retry_after_ms = 0
if allowed == 0 then
  if refill and refill > 0 then
    local missing = cost - new_tokens
    if missing < 0 then missing = 0 end
    retry_after_ms = math.floor((missing / refill) * 1000.0 + 0.5)
  else
    retry_after_ms = 0
  end
end

-- Persist state
redis.call('HMSET', key, 'tokens', new_tokens, 'ts', now_ms)

-- TTL: approx time to refill full capacity plus buffer
local ttl_sec
if refill and refill > 0 then
  ttl_sec = math.ceil(capacity / refill) + 5
else
  ttl_sec = 3600
end
redis.call('EXPIRE', key, ttl_sec)

-- Return: [allowed, tokens_remaining, capacity, retry_after_ms]
local remaining = math.floor(new_tokens)
return { allowed, remaining, capacity, retry_after_ms }
